CHANGES.md

```diff
@@ -66,6 +66,59 @@ The syntactic sugar for effect and record fields which lets one omit the `()` ha
 
 It is now possible to annotate type variables with `Mono` restriction, e.g. `sig id : (a::(Any,Mono)) -> a::(Any,Mono)`.
 
+### Recursive rows
+
+* We can now enter a recursive row effect type, such as `{ |(mu a.F:(() { |a}-> ()) {}-> b|c)}`.
+* **Breaking change**: recursive rows of different types cannot mix anymore - a variant can only contain fields with variant syntax (i.e. uppercase and separated by `|`) in its recursive row variable, and the same goes for others (e.g. records - lowecase, separated by `,`);
+  * these are correct: `[|(mu a. Foo:()|Bar:()|d)|]`, `(|(mu a. foo:(),bar:()|d))`
+  * these are not: `[|(mu a. foo:()|bar:()|d)|]`, `(|(mu a. foo:()|bar:()|d))`
+* **Breaking change**: Recursive variants with no directly exposed fields no longer require (nor allow) the vertical bar separating fields from the row variable: the following is no longer correct: `[| |(mu a . Foo:()|Bar:()|d)|]` (note the bar before `mu`)
```

Comment:
Hmm. I missed this point. I think we should add that syntax back in.

Reply:
I am not sure why I missed this, but I think it is "wrong" to be able to write `[|(mu a. ...)|]`, we should keep some syntactic consistency between record, effect, and variant rows.

Reply:
I think this was intentional at the time, to have a special sugar for variants, because `[|(mu r. φ)|]` is unambiguous, whereas `((mu r. φ))` is not the same as `(|(mu r. φ))`.

Ultimately there is a tradeoff to be made - either we want simplicity and uniformity (I think this is what I'd go for), or we want the syntax to be shorter even if it forces a special case to be made (this variant thing affords us to lose 2 characters: ` |`; there is a middle ground in allowing `[||` but then this may need to be a new lexeme).

Reply:
Yes. But we should not remove the ability to write `[| |p|]`, it is the unsugared/"correct" syntax.

Reply:
Ah, I see what you mean. I reckon this should only require duplicating one of the production rules for `vfields`, with the `vbar`.

Reply:
I have restored the syntax